[WIP] Add support for Ethereum

GildedHonour opened the pull request.

GildedHonour commented:
It's unclear wheather or not there's the plural form for Ethereum. I suppose not.

GildedHonour commented:
Finished yesterday. Now I'm testing it.

Vivek-Patil commented:
@GildedHonour Is this the full code? i cant see the address generation code here and all the rpc methods which are added in coinrpc for ethereum, from where are these methods called? 🤔 Am i missing something?

GildedHonour commented:
@Vivek-Patil 

> all the rpc methods which are added in coinrpc for ethereum
I've implemeted  the methods from https://github.com/ethereum/wiki/wiki/JSON-RPC  . What methods do you mean? 

If these https://github.com/rubykube/peatio/blob/devel/lib/coin_api.rb ones, I started working on this task before the new methods were added to coin_api.rb

> the address generation code
Can you elaborate?

yivo commented:
@GildedHonour Will it be possible for you to start rebasing ETH branch to devel branch after testing the current implementation?

Vivek-Patil commented:
@GildedHonour I got that the methods are implemented from wiki, but how these integrate with the peatio code? Like for example listtransactions method from coins.rb is called to get transactions. How is this used in eth ?

Vivek-Patil commented:
@GildedHonour I currently have the master peatio rubykube branch pulled on my local and i am able to run the whole code and use btc, xrp deposits and withdrawals.Am i missing something? Am i having the old code?

yivo commented:
@Vivek-Patil The latest code lives in devel branch.

GildedHonour commented:
@Vivek-Patil I don't know -- different coins have different set of PRC procuderues or some procedures doing the same thing but named differently. Therefore there still will be some distinctions in the APIs, even if we agree on some common interface or set of methods that exist in all the coins.

yivo commented:
@GildedHonour Have you started moving to devel branch?

GildedHonour commented:
@yivo not yet, I plan to do that today.

GildedHonour commented:
@yivo 
1) I've tested Ethereum RPC client using the npm utility "ethereumjs-testrpc" as a testnet. Some of the methods aren't implemented in it but the main ones are. Therefore, in a nushell, the client works. The futher testing will be needed.

2) I've squashed all the commits into a single one and rebased issue_334:

     git rebase devel

and commited it. There're the conflicts here. I have the branch "devel" locally but no conflicts. How can I properly pull the conflicts or the current version from here so that I can resolve them and commit?

GildedHonour commented:
And I'm ready to take a new task.

yivo commented:
@GildedHonour Can you please rebase to deven, fix conflicts and then pick https://github.com/rubykube/peatio/pull/437? @mkuvshynnikov pinged you.

GildedHonour commented:
@yivo 
Ok, merging.
 #437 -- it's not been merged into "devel" yet, how can I pick the commits from it?

GildedHonour commented:
@yivo lib/coin_rpc.rb was added, then deleted. It's used by, for instance, lib/coin_rpc/eth.rb and bch.rb
Do we need it?

GildedHonour commented:
3. And also I'll rebase #348 into this PR as Nick asked me

yivo commented:
@GildedHonour 

1. #348 is not correct PR, https://github.com/rubykube/peatio/pull/437 is
2. We refactored CoinRPC and renamed it to CoinAPI by adding single API interface for all currencies. You should check CoinAPI::Base, other implementations and create your one for ETH & ERC20 (you can grab some good code from #437, and you should grab specs from it).
3. Please, please rebase very carefully to DEVEL branch. We have added many changes.
4. Use test.bitgo.com for testing currencies since it is easy way.
5. We need to assign «gold master» to the status of this PR tomorrow, I will review it and we will merge it on weekends.

GildedHonour commented:
@yivo 
4. How should they be tested?

Also - test.bitgo.com  requires google authenticator to login and there's no way to skip this.

GildedHonour commented:
@yivo

All has been rebased and merged, removed the old stuff from server.en.yml  

One issue: the login button has disappeared for some reason

![screenshot_2018-02-10_09-23-07](https://user-images.githubusercontent.com/2032888/36057027-93ca3620-0e44-11e8-91ea-91af738d35b5.png)

GildedHonour commented:
![screenshot_2018-02-10_14-45-08](https://user-images.githubusercontent.com/2032888/36059332-8922d9ca-0e71-11e8-97a4-9a7b5fc6f64a.png)

GildedHonour commented:
and ws too:

![screenshot_2018-02-10_17-38-42](https://user-images.githubusercontent.com/2032888/36060687-51a7526a-0e89-11e8-8ccd-0f1dc8a47680.png)

yivo commented on app/controllers/admin/deposits/ethereum_controller.rb:
```diff
@@ -0,0 +1,21 @@
+module Admin
+  module Deposits
+    class EthereumController < ::Admin::Deposits::BaseController
```
@GildedHonour Controller is outdated.

yivo commented on app/controllers/admin/withdraws/ethereum_controller.rb:
```diff
@@ -0,0 +1,28 @@
+module Admin
+  module Withdraws
+    class EthereumController < ::Admin::Withdraws::BaseController
```
@GildedHonour Outdated code.

yivo commented on app/models/concerns/withdraws/coinable.rb:
```diff
@@ -30,4 +30,3 @@ def as_json(options={})
 
   end
 end
-
```
@GildedHonour Rollback changes pls.

yivo commented on app/views/admin/deposits/ethereum/index.html.erb:
```diff
@@ -0,0 +1,29 @@
+<%= table_for @ethereum, class: 'table table-condensed table-hover' do |t| %>
```
@GildedHonour Outdated.

yivo commented on app/views/admin/withdraws/ethereum/_table.html.erb:
```diff
@@ -0,0 +1,18 @@
+<%= table_for ethereum, class: 'table table-condensed table-hover', model: Withdraws::Ethereum do |t| %>
```
@GildedHonour Outdated.

yivo commented on app/views/admin/withdraws/ethereum/index.html.erb:
```diff
@@ -0,0 +1,20 @@
+<% unless @one_ethereum.empty? %>
```
@GildedHonour Outdated.

yivo commented on app/views/admin/withdraws/ethereum/show.html.erb:
```diff
@@ -0,0 +1,43 @@
+<div class="row">
```
@GildedHonour Outdated.

yivo commented on app/views/private/assets/_eth_assets.html.erb:
```diff
@@ -0,0 +1,50 @@
+<div class="panel panel-default">
```
@GildedHonour We removed this code.

yivo commented on app/views/private/withdraws/ethereum/edit.html.erb:
```diff
@@ -0,0 +1,16 @@
+<% defaults = {as: :display, required: false, hint: false} %>
```
@GildedHonour We removed this code.

yivo commented on app/views/private/withdraws/ethereum/new.html.erb:
```diff
@@ -0,0 +1,60 @@
+<%= simple_form_for @withdraw, as: :withdraw, scope: :ethereum do |f| %>
```
@GildedHonour Still outdated.

yivo commented on config/locales/client.en.yml:
```diff
@@ -169,3 +175,12 @@ en:
         done: Done
         canceled: Canceld
         failed: Failed
+
```
@GildedHonour What this is doing here?

yivo commented on lib/coin_rpc.rb:
```diff
@@ -0,0 +1,28 @@
+module CoinRPC
```
@GildedHonour Why do you have both CoinAPI & CoinRPC?

yivo commented on lib/coin_rpc/bch.rb:
```diff
@@ -0,0 +1,5 @@
+module CoinRPC
```
@GildedHonour Outdated.

yivo commented on lib/coin_rpc/eth.rb:
```diff
@@ -0,0 +1,322 @@
+class CoinRPC::ETH < CoinRPC::BaseRPC
```
@GildedHonour Outdated.

yivo commented on lib/coin_rpc/eth.rb:
```diff
@@ -0,0 +1,322 @@
+    send_post_request('eth_getBlockByHash', h, ret_val_as_objs)
+  end
+
+  def get_block_by_number(block_num_data, ret_val_as_objs)
```
@GildedHonour Don't use `get_` convention in Ruby.

GildedHonour replied:
1) What convention is that - not using "get_" ?

2) These methods mimic the names of the methods of Eth RPC and thus named accordingly, with camel_case.

yivo commented on lib/coin_rpc/eth.rb:
```diff
@@ -0,0 +1,322 @@
+    send_post_request('eth_getTransactionByBlockNumberAndIndex', block_num_data, idx)
+  end
+
+  def get_transaction_by_block_number_and_index(block_num_data, idx)
```
@GildedHonour Why do you have so many methods implemented?

GildedHonour replied:
These are the methods that exist in Eth RPC.

yivo commented on lib/coin_rpc/eth.rb:
```diff
@@ -0,0 +1,322 @@
+    send_post_request('web3_sha', data)
+  end
+
+  def net_version
```
@GildedHonour Why this ie needed in Peatio.

yivo commented on lib/coin_rpc/eth.rb:
```diff
@@ -0,0 +1,322 @@
+
+  # todo move to mixin or parent class
+  def send_post_request(name, *args)
+    http = Net::HTTP.new(@uri.host, @uri.port)
```
@GildedHonour Use Faraday.

yivo commented on lib/generators/config/templates/currencies.yml.erb:
```diff
@@ -94,6 +92,10 @@
   quick_withdraw_max: 10
   blockchain:         https://www.blocktrail.com/BCC/tx/#{txid}
   address_url:        https://www.blocktrail.com/BCC/address/#{address}
+  assets:
```
@GildedHonour Not needed.

GildedHonour replied:
where do the values of blockchain and address_url come from?

yivo commented on lib/generators/config/templates/currencies.yml.erb:
```diff
@@ -65,8 +65,6 @@
   #
   # The rest of configuration.
   quick_withdraw_max: 10000
-  blockchain:         https://bithomp.com/explorer/#{txid}
```
@GildedHonour Why this is removed?

yivo commented on lib/generators/config/templates/currencies.yml.erb:
```diff
@@ -140,4 +142,25 @@
   blockchain:         https://live.blockcypher.com/dash/tx/#{txid}
   address_url:        https://live.blockcypher.com/dash/address/#{address}
+
+- id: 7
```
@GildedHonour Outdated configuration.

yivo commented on lib/generators/config/templates/currencies.yml.erb:
```diff
@@ -140,4 +142,25 @@
+- id: 7
+  key: ethereum
+  code: eth
+  symbol: "Ξ"
```
@GildedHonour Use single quotes.

yivo commented on lib/generators/config/templates/currencies.yml.erb:
```diff
@@ -140,4 +142,25 @@
+  json_rpc_endpoint: http://user:password@127.0.0.1:8545
+  hot_wallet_address: ~
+
+- id: 8
```
@GildedHonour Outdated configuration.

yivo commented on lib/generators/config/templates/markets.yml.erb:
```diff
@@ -129,4 +129,31 @@
+  ask: { fee: 0.0015, currency: eth, fixed: 4 }
+
+
+  # todo: add also: eth/ltc, eth/dash, etc...
```
@GildedHonour Agreed.

yivo replied:
@GildedHonour Any progress?

yivo commented on spec/fixtures/currencies.yml:
```diff
@@ -16,3 +16,21 @@
   key: protoshare
   code: pts
   coin: true
+- id: 7
```
@GildedHonour Outdated configs.

yivo commented:
@GildedHonour Please recheck https://github.com/rubykube/peatio/pull/381#issuecomment-364062755

yivo commented:
@GildedHonour Login works well. Check your configs.

yivo commented on lib/coin_api/eth.rb:
```diff
@@ -0,0 +1,124 @@
+    end
+
+    # You must pass password to create ethereum address
+    def create_address!(password)
```
@GildedHonour It should not take argument. You can generate some random password and return { address: '12345', secret: password }.

GildedHonour replied:
Where does the value  '12345' of address come from?

yivo replied:
@GildedHonour You can generate password by yourself and return it's value assigned to `secret` while `address` is the address ETH should generate for you according to secret.

Example return value: `{address: 'rnYtF9YjLV2dbqQS4JFN9YRSxPQEWbMGWC', secret: 'LuohKDoZEq9u4Ysy7V5uHhrSmG5KspTE'}` for XRP.

GildedHonour replied:
@yivo 
The client calls `CoinAPI::ETH.new.create_address!(...)` and expects to get `address` and perhaps password and secret. What do you mean "generate by myself"? Is address a completely random value or length N? How's `secret` generated? Are there examples?

yivo replied:
@GildedHonour You can generate some secret using ruby.

GildedHonour replied:
@yivo I didn't create `lib/coin_api/eth.rb`. I created another Eth class which already has been deleted because of the merge. 

And my questions aren't about the concrete implemention but rather interface of the class `lib/coin_api/eth.rb `. Why isn't `password` needed, for instance? Whom can I discuss that with?

yivo replied:
@GildedHonour Because API is designed so. It doesn't accept argument.

GildedHonour replied:
I'll fix this today.

yivo commented on lib/generators/config/templates/currencies.yml.erb:
```diff
@@ -140,4 +140,35 @@
+# todo
+  blockchain:         https://bithomp.com/explorer/#{txid}
+  address_url:        https://bithomp.com/explorer/#{address}
+
```
@GildedHonour please update the configs according to new format (with docs etc).

`blockchain` is used for exploring transaction details
`address_url` – for exploring address in blockchain
you can google some

GildedHonour replied:
`ethereum blockchain` and `ethereum blockchain url` returns nothing usefull. Neither `erc20` does. Please elaborate.

yivo commented on spec/fixtures/currencies.yml:
```diff
@@ -16,3 +16,21 @@
+  quick_withdraw_max: 1
+  json_rpc_endpoint: http://127.0.0.1:8545
+  token_address: '0x6a472aab762eabd632818d07d3c46b4bca6ae733'
+  hot_wallet_address: '0x89Af4bF02126b56fc2c24BC324154fF3628Bd946'
```
@GildedHonour pls document what are these two in currencies.yml.erb

GildedHonour replied:
token_address and hot_wallet_address don't exist in currencies.yml.erb

However, this in particular was added by Vadim Sh. Should I remove   token_address and hot_wallet_address?

yivo replied:
@GildedHonour I don't know what the are used for. I am not ETH expert. Can you check it?

GildedHonour replied:
@vshatravenko -- Are token_address and hot_wallet_address needed given that they aren't used in currencies.yml.erb? Can they be deleted?

yivo replied:
@GildedHonour Again: I am not ETH expert. I don't know. Peatio should work 100% after your PR (deposit / withdraw / trading). If these configs are required you can keep them, if not – you can delete, or modify. But write docs for them.

GildedHonour replied:
@yivo my question was addressed to @vshatravenko  I didn't add those keys. They're used anywhere, it seems, but vshatravenko might've had an intention for adding them -- I don't know.

GildedHonour commented:
@yivo I've made the requested changes.  Since the base has been changed from devel to master by @mod which will require some time to merge,  should I merge the changes with master or switch back to devel?

rraallvv commented:
It seems branch `1-8-stable` has support for ETH, although I can't find any information on how to add the function to be notified of received transactions, just like Bitcoin's -walletnotify.
Could someone tell me if that function is documented somewhere.

GildedHonour commented:
Some idiots managing the development process on the top couldn't even remember what they were saying and what I was saying after a few days have gone, when I was working on this. Let alone doing programming -- too difficult.
